File: SinglyLL/singly_linked_list.js
// peice of data - val
// reference to next node - next

class Node {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

// Not the best way to implement
// let first = new Node("Cam");
// first.next = new Node("Adam");
// first.next.next = new Node("Sam");
// first.next.next.next = new Node("Rick");

class SinglyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    push(val) {
        let node = new Node(val);

        if (this.head === null) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail.next = node;
            this.tail = node;
        }
        this.length++;
        return this;
    }
    pop() {
        if (!this.length) {
            return undefined;
        }
        let current = this.head;
        let newTail = current;

        while (current.next) {
            newTail = current;
            current = current.next;
        }
        this.tail = newTail;
        this.tail.next = null;
        this.length--;

        if (this.length === 0) {
            this.head = null;
            this.tail = null;
        }
        return current;
    }
    shift() {
        if (!this.length) {
            return undefined;
        }
        let currentHead = this.head;
        this.head = currentHead.next;
        this.length--;
        if (this.length === 0) {
            this.tail = null;
        }
        return currentHead;
    }
    unshift(val) {
        let newHead = new Node(val);
        if (!this.head) {
            this.head = newHead;
            this.tail = this.head;
        } else {
            newHead.next = this.head;
            this.head = newHead;
        }
        this.length++;
        return this;
    }
    get(index) {
        if (index < 0 || index >= this.length) {
            return null;
        }
        let count = 0;
        let current = this.head;
        while (count < index) {
            current = current.next;
            count++;
        }
        return current;
    }
    set(index, val) {
        let node = this.get(index);
        if (!node) {
            return false;
        }
        node.val = val;
        return true;
    }
}

let list = new SinglyLinkedList();

File: SinglyLL/Python/singly_linked_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

    def __str__(self):
        return f'Node {self.val}'


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def __str__(self):
        return f'Head: {self.head}. Tail: {self.tail}. Size: {self.size}'

    def push(self, val):

        new_node = Node(val)

        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.size += 1
        return self

    def pop(self):

        if not self.head:
            return None

        current = self.head
        new_tail = current

        while current.next:
            new_tail = current
            current = current.next

        self.tail = new_tail
        self.tail.next = None
        self.size -= 1

        if self.size == 0:
            self.head = None
            self.tail = None

        return current

    def shift(self):

        if not self.head:
            return None

        old_head = self.head
        self.head = old_head.next
        self.size -= 1

        if self.size == 0:
            self.tail = None

        return old_head

    def unshift(self, val):

        new_node = Node(val)

        if not self.head:
            self.head = new_node
            self.tail = new_node

        new_node.next = self.head
        self.head = new_node

        self.size += 1

        return self


sll = SinglyLinkedList()

sll.push(2)
print(sll.unshift(3))


# curr = sll.head
# while curr:
#     print(curr)
#     curr = curr.next
